### PR TITLE
Update reference, MiMa previous version and sync TASTy version

### LINKED ISSUE #13447: Verify TASTy versions before releases
We need to verify before releases that TASTy is set to a correct version.

Here is a very simplistic approach that could be automated to ensure that
we don't accidentally release experimental TASTy in a stable release,
or stable TASTy in a Nightly release:

(`mStable` refers to the `TastyFormat.MinorVersion` of the most recent stable Scala release)

```python
if MajorBump:
  raise "are you really sure?"
else if Nightly or (MinorBump and RC or Milestone):
  assert TastyFormat.ExperimentalVersion > 0
  assert TastyFormat.MinorVersion > mStable
else: # release is final
  assert ExperimentalVersion == 0
  if PatchBump:
    assert TastyFormat.MinorVersion == mStable
  else if MinorBump:
    assert TastyFormat.MinorVersion == mStable + 1
```

with the proposal above, the only clear weak link is that
`TastyFormat.MinorVersion` should be increased in a nightly release
before releasing the next Minor, but I propose that this is
difficult to miss as it should be clear that
merging a `needs-minor-release` PR needs a TASTy version bump

A next step would be to automate the weak-link I described above.

A simple way could be automating the creation of an issue whenever
a `needs-minor-release` PR is merged, otherwise it will probably
be too complex to analyse automatically (
  requires analysing TreePickler, TastyFormat, API additions, Code synthesis changes, more)

### PR DESCRIPTION
**Pull request by WojciechMazur**
* Update reference version to 3.6.3-RC1 (from 3.6.0)
* Update mima previous binary verison to 3.6.2 (instead of unofficial 3.6.1) 
* Set TASTy version to `28.7-experimental-1` - it should have been set when branching of 3.6.3. 
* We now document better how and when tasty version should be set 
* Add additional runtime test to ensure we don't emit invalid TASTy version during Release / NIGHTLY releases and the expected version set in build matches version defined in TastyFormat - resolves #13447
 
Reference version and TASTy version bump needs to be performed at the same time (otherwise usage of compiler with stable TASTY would lead to failure in from-tasty non-bootstrapped tests)

[test_non_bootstrapped]

### CODE REVIEW

File: tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        case s"$major.$minor-experimental-$experimental" => TastyVersion(major.toInt, minor.toInt, experimental.toInt)
+        case s"$major.$minor" if minor.forall(_.isDigit) => TastyVersion(major.toInt, minor.toInt, 0)
+        case other => fail(s"Invalid TASTY version string: $other")
+      }
+    assertEquals(CurrentTastyVersion, expectedVersion)
+  }
+  
+  // Tested only in nightly / release builds
+  // Protects from publishing artifacts with incorrect TASTY version
+  @Test def testReleasedTastyVersion(): Unit = {
```
Comment (WojciechMazur) [inline tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala:30]: This test actually would be never executed - RELEASEBUILD and NIGHTLYBUILD are set only for the release but not in tests executed as a requirement to start releasing job

**Comment by bishabosha** [comment]
glad there's finally a real test :)

File: tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  
+  val CurrentTastyVersion = TastyVersion(TastyFormat.MajorVersion, TastyFormat.MinorVersion, TastyFormat.ExperimentalVersion)
+  
+  // Needs to be defined in build Test/envVars
+  val ExpectedTastyVersionEnvVar = "EXPECTED_TASTY_VERSION"
+  
+  @Test def testBuildTastyVersion(): Unit = {
+    val expectedVersion = sys.env.get(ExpectedTastyVersionEnvVar)
+      .getOrElse(fail(s"Env variable $ExpectedTastyVersionEnvVar not defined"))
+      .match {
```
Comment (hamzaremmal) [inline tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala:19]: Does this test even run? This is not valid syntax...

Comment (prolativ) [reply tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala:19]: `foo.match {...}` is a valid alternative syntax for `foo match {...}` in scala 3

Comment (hamzaremmal) [reply tasty/test/dotty/tools/tasty/BuildTastyVersionTest.scala:19]: The metals interpret the code as Scala 2 for some reason:

Screenshot 2025-01-29 at 12 06 34

**Pull request merged**